src/main/scala/io/iohk/ethereum/jsonrpc/FilterManager.scala

```diff
@@ -0,0 +1,271 @@
+    val lastCheckBlock = lastCheckBlocks.getOrElse(id, bestBlockNumber)
+
+    val filterOpt = filters.get(id)
+    filterOpt.foreach { _ => lastCheckBlocks += (id -> bestBlockNumber) }
+    resetTimeout(id)
+
+    filterOpt match {
+      case Some(logFilter: LogFilter) =>
+        sender() ! LogFilterChanges(getLogs(logFilter, Some(lastCheckBlock + 1)))
+
+      case Some(_: BlockFilter) =>
+        sender() ! BlockFilterChanges(getBlockHashesAfter(lastCheckBlock).takeRight(maxBlockHashesChanges))
+
+      case Some(_: PendingTransactionFilter) =>
+        getPendingTransactions().map { pendingTransactions => // TODO: should this return only pending transactions added since last poll?
+          PendingTransactionFilterChanges(pendingTransactions.map(_.hash))
+        }.pipeTo(sender())
+
+      case None =>
+        sender() ! LogFilterChanges(Nil)
+    }
+  }
+
+  private def getLogsFromBlock(filter: LogFilter, blockHeader: BlockHeader, receipts: Seq[Receipt]): Seq[Log] = {
+    val bytesToCheckInBloomFilter = filter.address.map(a => Seq(a.bytes)).getOrElse(Nil) ++ filter.topics.flatten
+
+    receipts.zipWithIndex.foldLeft(Seq[Log]()) { case (logsSoFar, (receipt, txIndex)) =>
+      if (BloomFilter.containsAnyOf(receipt.logsBloomFilter, bytesToCheckInBloomFilter)) {
+        logsSoFar ++ receipt.logs.zipWithIndex
+        .filter { case (log, _) => filter.address.forall(_ == log.loggerAddress) && topicsMatch(log.logTopics, filter.topics) }
+        .map { case (log, logIndex) =>
+          val blockBody = blockchain.getBlockBodyByHash(blockHeader.hash).get
+          val tx = blockBody.transactionList(txIndex)
+          Log(
+            logIndex = logIndex,
+            transactionIndex = Some(txIndex), // TODO: None is this is a pending block (?)
+            transactionHash = Some(tx.hash), // TODO: None is this is a pending block (?)
+            blockHash = blockHeader.hash,
+            blockNumber = blockHeader.number,
+            address = log.loggerAddress,
+            data = log.data,
+            topics = log.logTopics)
+        }
+      } else logsSoFar
+    }
+  }
+
+  private def topicsMatch(logTopics: Seq[ByteString], filterTopics: Seq[Seq[ByteString]]): Boolean = {
+    (filterTopics zip logTopics).forall { case (filter, log) => filter.isEmpty || filter.contains(log) }
+  }
+
+  private def getBlockHashesAfter(blockNumber: BigInt): Seq[ByteString] = {
+    val bestBlock = appStateStorage.getBestBlockNumber()
+
+    @tailrec
+    def recur(currentBlockNumber: BigInt, hashesSoFar: Seq[ByteString]): Seq[ByteString] = {
+      if (currentBlockNumber > bestBlock) {
+        hashesSoFar
+      } else blockchain.getBlockHeaderByNumber(currentBlockNumber) match {
+        case Some(header) => recur(currentBlockNumber + 1, hashesSoFar :+ header.hash)
+        case None => hashesSoFar
+      }
+    }
+
+    recur(blockNumber + 1, Nil)
+  }
+
+  private def getPendingTransactions(): Future[Seq[SignedTransaction]] = {
```

Comment:
It seems this one doesn't work as parity does. Our implementation is returning all the pending transactions every time I query them. In parity, it only returns the ones It hasn't returned before.

This doc statement is confusing, but might be the reason _Creates a filter in the node, to notify when new pending transactions arrive_